### ISSUE TITLE
weird behavior with non-broadcasting masked_select: a bug?

### ISSUE THREAD

**Issue by aosokin** [pytorch/pytorch#2213]
Hi, when I do this
```
import torch
A = torch.rand(5, 5, 1)
mask = torch.ByteTensor(5, 5).fill_(1)
B = torch.masked_select(A, mask)
B.size()
```
I get a warning about no broadcasting, but the output tensor is of weird size of `torch.Size([125])`. Its the same of both CPU and a GPU.
I'm on the head on the master branch.

It this an expected behavior or a bug? Some of older code broke because of this.

**Comment by bheinzerling** [comment]
output has size `torch.Size([25])` with pytorch version 0.1.12_2

**Comment by aosokin** [comment]
@bheinzerling this is connected to the new broadcasting behavior, which was not in 0.1.12
The warning says that the system returns to the old behavior, but it does not

**Comment by gchanan** [comment]
@aosokin can you show the warning you get, as well as any warning flags you enabled (e.g. torch.utils.backcompat.broadcast_warning=True)?

To answer your later question, this behavior changed with the introduction of broadcasting; "A" and "mask" are being broadcasted to shape (5,5,5) and since the mask is all ones you get an output tensor of size 5*5*5.

**Comment by aosokin** [comment]
@gchanan with this test the warning is gone, sorry. But, with this
```
import torch
A = torch.rand(78, 64, 1)
mask = torch.ByteTensor(78, 64).fill_(1)
B = torch.masked_select(A, mask)
print(B.size())
```
I get warning
```
tmp.py:4: UserWarning: self and mask not broadcastable, but have the same number of elements.  Falling back to deprecated pointwise behavior.
  B = torch.masked_select(A, mask)
```
and the results is of size `torch.Size([4992])`, which is without broadcasting.
I did not turn on any special flags.

I guess the discrepancy between the two is a bug.

**Comment by gchanan** [comment]
It's not a bug, in the second case the tensors are not broadcastable, as the message says.  In the first case they are.  See here: http://pytorch.org/docs/master/notes/broadcasting.html

**Comment by aosokin** [comment]
@gchanan Thanks for the explanation, I got confused how broadcasting is defined. Sorry!
